renderdoc/driver/vulkan/vk_acceleration_structure.cpp

```diff
@@ -450,320 +618,261 @@ void VulkanAccelerationStructureManager::CopyAccelerationStructure(
+      CHECK_VKR(m_pDriver, vkr);
 
-  // not using SERIALISE_ELEMENT_ARRAY so we can deliberately avoid allocation - we serialise
-  // directly into upload memory
-  ser.Serialise("Serialised AS"_lit, contents, contentsSize, SerialiserFlags::NoFlags).Important();
+      VkCommandBuffer cmd = m_pDriver->GetInitStateCmd();
+      if(cmd == VK_NULL_HANDLE)
+      {
+        RDCERR("Couldn't acquire command buffer");
+        return false;
+      }
 
-  // unmap the resource we mapped before - we need to do this on read and on write.
-  bool isTLAS = false;
-  if(!IsStructuredExporting(state) && mappedMem.mem != VK_NULL_HANDLE)
-  {
-    if(replayingAndReading)
-    {
-      // first ensure we flush the writes from the cpu to gpu memory
-      const VkMappedMemoryRange range = {
-          VK_STRUCTURE_TYPE_MAPPED_MEMORY_RANGE,        NULL, Unwrap(mappedMem.mem), mappedMem.offs,
-          AlignUp(mappedMem.size, nonCoherentAtomSize),
+      VkBufferCopy region = {
+          0,
+          0,
+          AlignUp(geomData.memSize, nonCoherentAtomSize),
       };
+      ObjDisp(d)->CmdCopyBuffer(Unwrap(cmd), uploadMemory.buf, Unwrap(gpuBuf), 1, &region);
 
-      vkr = ObjDisp(d)->FlushMappedMemoryRanges(Unwrap(d), 1, &range);
-      CHECK_VKR(m_pDriver, vkr);
-
-      // Read the AS's BLAS handle count to determine if it's top or bottom level
-      isTLAS = *((uint64_t *)(contents + handleCountOffset)) > 0;
+      geomData.replayBuf = gpuBuf;
     }
-
-    ObjDisp(d)->UnmapMemory(Unwrap(d), Unwrap(mappedMem.mem));
   }
 
   SERIALISE_CHECK_READ_ERRORS();
 
-  if(IsReplayMode(state) && contentsSize > 0)
+  if(IsReplayMode(state))
   {
-    VkInitialContents initialContents(eResAccelerationStructureKHR, uploadMemory);
-    initialContents.isTLAS = isTLAS;
-    initialContents.buf = uploadBuf;
+    VkInitialContents initialContents;
+    initialContents.type = eResAccelerationStructureKHR;
+    initialContents.accelerationStructureInfo = asInfo;
 
     m_pDriver->GetResourceManager()->SetInitialContents(id, initialContents);
   }
+  else
+  {
+    asInfo->Release();
+  }
 
   return true;
 }
 
-template bool VulkanAccelerationStructureManager::Serialise(ReadSerialiser &ser, ResourceId id,
-                                                            const VkInitialContents *initial,
-                                                            CaptureState state);
-template bool VulkanAccelerationStructureManager::Serialise(WriteSerialiser &ser, ResourceId id,
-                                                            const VkInitialContents *initial,
-                                                            CaptureState state);
-
 void VulkanAccelerationStructureManager::Apply(ResourceId id, const VkInitialContents &initial)
 {
-  VkCommandBuffer cmd = m_pDriver->GetInitStateCmd();
-  if(cmd == VK_NULL_HANDLE)
-  {
-    RDCERR("Couldn't acquire command buffer");
-    return;
-  }
+  const VkAccelerationStructureKHR wrappedAS =
```

Comment:
I found that repeated AS builds had a pretty big impact on replay time, so I ended up for BLASs building into a temporary scratch GPU-local AS of my own which can then just be copied into the resource being apply'd. The copy ends up much faster than a rebuild.

If you rebase onto latest you'll find that `Apply_InitState` now takes a non-const `VkInitialContents` to allow for this caching and update-in-place.

This would also let you release the uploaded build inputs too (at least for now, in future if we need them for debugging we can keep it around), and in theory the scratch buffer too.

Reply:
> I ended up for BLASs building into a temporary scratch GPU-local AS of my own

What's the best way of creating this AS?  It needs to be identical to the 'real' AS but AFAICT the creation info has been lost - do I need to capture all the creation info in `VkAccelerationStructureInfo` too or is there a simpler way?

Reply:
Actually looking at `VkAccelerationStructureCreateInfoKHR` , the only part I don't have is whatever is in the `pNext` chain specificallty `VkOpaqueCaptureDescriptorDataCreateInfoEXT`.  No idea if that's a problem.